### ISSUE TITLE
[Request] Themes for areas

### ISSUE THREAD

**Issue by Heraes-git** [sojimaxi/grav-plugin-widget#17]
# Actual paradigm

For now, we can call a widget area directly in our template, like this : 
```
{{ widget('left') }}
{% block content %}
    {# Dynamic content of the page #}
{% endblock -%}
{{ widget('right') }}
```
If a widget area loads, it will get the widgets, wich themselves will use the modular templates defined in each widget, in order to build them.

It is fine, **but if we want to structure the page in order to create columns to store the widgets on the sides of the content, we have to add some HTML and CSS elements** in order to wrap the widgets generated. Like, for instance : 
```
<div id="widgets_left">{{ widget('left') }}</div>
{% block content %}
    {# Dynamic content of the page #}
{% endblock -%}
<div id="widgets_right">{{ widget('right') }}</div>
```
This leading to an undesired consequence : **the `<div>` around the widgets will stay, wether any widget has loaded or not**, this creating a problem : the sizes dedicated to those `<div>` (margin, padding) will take a useless **space** on the page, and will create blank areas.

So the user is forced to avoid HTML wrapping, and is stuck in a basic linear injection of widgets on his page. Or he has to use JavaScript and to hide elements when no inner HTML code has been generated. Or, in a last resort, to trick with the widgets' templates, but in a very limited way.

# Solution 

For solving this, **the call of widgets areas should trigger some dedicated files**, embedding HTML and CSS capability, that the user should be able to customize freely, by overloading them in `/user/data/widget`, **or (even better) directly in the theme's template folder.**

# Ideas on how to proceed

Widget areas should still work on the same basis, and be called like actually : 

```
{{ widget('area_name') }}
```

But instead of directly processing the widgets generation, it should be connected to a .html.twig file _(prefixed with the area name)_, embedding the HTML, CSS, and **Twig call of the widgets**, like for instance :
```
<div>
{{ widgets }}
</div>
```
This last twig call should target the actual plugin process of generating the widgets.

## Place to store the areas' templates

I suggest to : 
- Natively integrate a default template for areas, in the plugin's `templates` folder.
- To modify the YAML areas definition in order to integrate a template location : 
```
widget:
  areas: 
      - name: top
        template: '/partials/widget/top'
      - name: bottom
        template: '/partials/widget/bottom'
      - etc...
```
- And to let the user to create those files. In the example above, I used `/partials/widget` as folder, so respectively to this example, the user would create two files named `bottom.html.twig` and `top.html.twig` in `/user/themes/theme_name/templates/partials/widget`. But he could have chosen some different names for the files.

## Simplification 

A simpler way to design this YAML and template structure would be : 

```
widget:
  areas: [top,bottom]
  templates: '/partials/widget'
```
But it would force the user to name the areas' templates with the exact same name than the areas themselves, **and would avoid him to reuse same templates for different areas**.

# In conclusion

The idea is to add an intermediary step, in wich some HTML elements would be carried with widget areas, this to give the plugin the ability to transform pages structure.

`Widget` is a fantastic plugin, offering people the power to mimic Wordpress and other CMS with brick-composition features : but because of the actual HTML wrapping configuration, the result is pretty rigid and disappointing.

Adding an additional template layer would definitely bring the plugin to its full potential.

**Comment by Heraes-git** [comment]
Well, that's very unfortunate for the time I spent on writting this request, but I've just found that just by controlling the widget() variable : 
```
{% if widget('left') %}<div id="widgets_left">{{ widget('left') }}</div>{% endif %}
```
we can skip the HTML element generation if the area is empty.

So, you can close this thread. Or if you want to add a template feature, you're welcome, but I think it's useless, given that we can use `{% block %}` or `{% include %}` in order to satisfy any template separation, based on the same _variable control_ that I showed above.

**Comment by sojimaxi** [comment]
Thanks @Romarain. Your last [comment](https://github.com/sojimaxi/grav-plugin-widget/issues/17#issuecomment-476900703) sums it up.
